### PR TITLE
955 - Download Config Validation for Batch Jobs

### PR DESCRIPTION
**Pull request by avrohomgottlieb**
## Issue Number

Closes https://github.com/AlexsLemonade/scpca-portal/issues/955

## Purpose/Implementation Notes

This PR removes the submitting of invalid jobs to Batch. 

We had originally just submitted all project (or sample) download_config combinations to Batch, and let the invalid ones return early. However, spinning up Batch instances with jobs that will not create a computed file is not an optimal solution (and is not cost effective). As such, this PR tightens up `dispatch_to_batch` by first validating which download configs produce libraries with projects and samples and only then sending them through. 

Changes are also applied to `loader::generate_computed_files` to maintain consistency among the `load_data` and `dispatch_to_batch`+`generate_computed_file` workflows.

The patches in `test_loader` had to be slightly refactored in order to accommodate for this change as well. When the dust settled, the constants `GENERATED_PROJECT_DOWNLOAD_CONFIGS` and `GENERATED_SAMPLE_DOWNLOAD_CONFIGS` were removed because they were no longer necessary.

## Types of changes

What types of changes does your code introduce?

- New feature (non-breaking change which adds functionality)

## Functional tests

List out the functional tests you've completed to verify your changes work locally.

## Checklist

- [x] Lint and unit tests pass locally with my changes

## Screenshots

N/A

### CODE REVIEW

**Comment by avrohomgottlieb** [comment]
Because this PR touches `Library::get_project_libraries_from_download_config` and `Library::get_sample_libraries_from_download_config`, I'd like to take the opportunity to move the first method to the `Project` model and change the name to `Project::get_libraries`, and the same for the second method, and rename it `Sample::get_libraries`.

File: api/scpca_portal/models/project.py [hunk 1]
```diff
@@ -269,6 +269,13 @@ def get_download_config_file_paths(self, download_config: Dict) -> List[Path]:
             file_path for file_path in data_file_path_objects if file_path.parent.name != "merged"
         ]
 
+    def get_valid_download_configs(self) -> List[str]:
```
Comment (davidsmejia) [inline api/scpca_portal/models/project.py:272]: Let's break this up into two different properties `project.valid_download_configs` and `project.valid_download_config_names`

**Comment by davidsmejia** [comment]
> Because this PR touches `Library::get_project_libraries_from_download_config` and `Library::get_sample_libraries_from_download_config`, I'd like to take the opportunity to move the first method to the `Project` model and change the name to `Project::get_libraries`, and the same for the second method, renaming it `Sample::get_libraries`.

This sounds good to me, we should make download_config optional then and return all libraries if not passed.

File: api/scpca_portal/models/computed_file.py [hunk 2]
```diff
@@ -225,10 +227,12 @@ def get_sample_file(cls, sample, download_config: Dict) -> Self | None:
...
         creates a ComputedFile object which it then saves to the db.
         """
-        libraries = Library.get_sample_libraries_from_download_config(sample, download_config)
-        # If the query returns empty, then an error occurred, and we should abort early
+        libraries = sample.get_libraries(download_config)
+        # If the query returns empty, then throw an error occurred.
         if not libraries.exists():
-            return
+            raise ValueError(
+                "Invalid request: no libraries exist with this sample-download_config combination."
```
Comment (davidsmejia) [inline api/scpca_portal/models/computed_file.py:234]:
```suggestion
                "Unable to find libraries for download_config."
```

File: api/scpca_portal/models/computed_file.py [hunk 3]
```diff
@@ -151,10 +151,12 @@ def get_project_file(cls, project, download_config: Dict) -> Self | None:
...
         creates a ComputedFile object which it then saves to the db.
         """
-        libraries = Library.get_project_libraries_from_download_config(project, download_config)
-        # If the query returns empty, then an error occurred, and we should abort early
+        libraries = project.get_libraries(download_config)
+        # If the query returns empty, then throw an error occurred.
         if not libraries.exists():
-            return
+            raise ValueError(
+                "Invalid request: no libraries exist with this project-download_config combination."
```
Comment (davidsmejia) [inline api/scpca_portal/models/computed_file.py:158]:
```suggestion
                "Unable to find libraries for download_config."
```

File: api/scpca_portal/models/project.py [hunk 4]
```diff
@@ -185,6 +201,44 @@ def get_metadata(self) -> Dict:
...
 
+    def get_libraries(self, download_config: Dict = {}):  # -> QuerySet[Library]:
+        """
+        Return all of a project's associated libraries filtered by the passed download config.
+        """
+        if not download_config or download_config.get("metadata_only"):
+            return self.libraries.all()
+
+        if download_config not in common.PROJECT_DOWNLOAD_CONFIGS.values():
+            raise ValueError("Invalid download configuration passed. Unable to retrieve libraries.")
```
Comment (davidsmejia) [inline api/scpca_portal/models/project.py:212]:
```suggestion
            raise ValueError("Invalid download_config passed. Unable to retrieve libraries.")
```

File: api/scpca_portal/models/sample.py [hunk 5]
```diff
@@ -121,6 +137,21 @@ def get_metadata(self) -> Dict:
...
 
+    def get_libraries(self, download_config: Dict = {}):  # -> QuerySet[Library]:
+        """
+        Return all of a sample's associated libraries filtered by the passed download config.
+        """
+        if not download_config:
+            return self.libraries.all()
+
+        if download_config not in common.SAMPLE_DOWNLOAD_CONFIGS.values():
+            raise ValueError("Invalid download configuration passed. Unable to retrieve libraries.")
```
Comment (davidsmejia) [inline api/scpca_portal/models/sample.py:148]:
```suggestion
            raise ValueError("Invalid download_config passed. Unable to retrieve libraries.")
```

File: api/scpca_portal/test/models/test_project.py
Comment (davidsmejia) [inline api/scpca_portal/test/models/test_project.py:1]: Let's replace the download config implementations with references to common here.

**Review by davidsmejia — APPROVED** [review]

**Pull request merged**